### PR TITLE
🌱  Add make commands for CAPI operator

### LINKED ISSUE #4122: Create make commands for the operator
**User Story**

As a CAPI developer, I would like to have convenient make commands to build the operator, generate manifests, and build docker images.

**Detailed Description**
Some of the commands are below but not limited to
```
# Build the binary 
make operator 
# Generate go related go
make generate-go-operator
# Generate manifests for the operator
make generate-operator-manifests
# Build and push docker images
make docker-build-operator
make docker-push-operator-manifest

# Modify the following
make release
make release-manifests
```

/kind feature
/area api
/area operator

### PR DESCRIPTION
**Pull request by wfernandes**
**What this PR does / why we need it**:
This PR adds a bunch of make commands for the CAPI operator. It is built on top of the PR #4117. 
This PR should be merged after #4117.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4122
Ref #3833

### CODE REVIEW

**Comment by wfernandes** [comment]
/hold
Putting this on hold because this PR should be merged in after #4117.

**Comment by vincepri** [comment]
What do you think about having a separate Makefile for the operator in the relative folder and use `$(MAKE) -c /path/to/operator` to run commands in those?

We should probably do the same for the bootstrap, control plane providers, etc

**Comment by wfernandes** [comment]
Sure. I was just being consistent with what we have now.

I don't mind breaking it up into several Makefiles but can we do that as part of a separate issue. I'm happy to create the issue describing Makefiles for each CAPI component that makes sense.
WDYT?

**Comment by alexander-demichev** [comment]
I updated the PR, now it's based on operator code located in `exp`. 
@vincepri I agree with having a separate Makefile, this should make CAPI one more readable, especially for newcomers. Can this be tracked in a separate issue?

**Comment by alexander-demichev** [comment]
Some changes here address points from https://github.com/kubernetes-sigs/cluster-api/issues/4246

File: exp/operator/controllers/coreprovider_controller.go [hunk 1]
```diff
@@ -30,17 +29,15 @@ import (
...
 // +kubebuilder:rbac:groups=operator.cluster.x-k8s.io,resources=coreproviders,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operator.cluster.x-k8s.io,resources=coreproviders/status,verbs=get;update;patch
 
-func (r *CoreProviderReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("coreprovider", req.NamespacedName)
-
+func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	_ = log
```
Comment (MarcelMue) [inline exp/operator/controllers/coreprovider_controller.go:40]:
```suggestion
```

I am not sure why we need the logger in this PR. Leaving it as a noop seems fine to me.

File: exp/operator/config/default/manager_image_patch.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: controller-manager
+  namespace: system
+spec:
+  template:
+    spec:
+      containers:
+        - image: gcr.io/openshift-gce-devel/operator-controller-amd64:dev
```
Comment (MarcelMue) [inline exp/operator/config/default/manager_image_patch.yaml:10]: Why is this coming from an openshift image repo? I would have expected something like:

```
gcr.io/k8s-staging-cluster-api/cluster-api-operator:operator-0.4
```

File: exp/operator/config/default/kustomization.yaml [hunk 3]
```diff
@@ -1,36 +1,39 @@
...
 # Note that it should also match with the prefix (text before '-') of the namespace
 # field above.
-namePrefix: operator-
+namePrefix: capi-operator-
 
 # Labels to add to all resources and selectors.
-#commonLabels:
-#  someName: someValue
+commonLabels:
+  clusterctl.cluster.x-k8s.io/core: "capi-operator"
```
Comment (MarcelMue) [inline exp/operator/config/default/kustomization.yaml:13]: Why is it `clusterctl.cluster.x-k8s.io`? Did we agree on this somewhere? It seems a bit odd to me to have this clusterctl label key in the operator labels.

File: exp/operator/config/certmanager/certificate.yaml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # The following manifests contain a self-signed issuer CR and a certificate CR.
 # More document can be found at https://docs.cert-manager.io
-# WARNING: Targets CertManager 0.11 check https://docs.cert-manager.io/en/latest/tasks/upgrading/index.html for 
+# WARNING: Targets CertManager 0.11 check https://docs.cert-manager.io/en/latest/tasks/upgrading/index.html for
 # breaking changes
 apiVersion: cert-manager.io/v1alpha2
```
Comment (MarcelMue) [inline exp/operator/config/certmanager/certificate.yaml:5]: We should switch these to `cert-manager.io/v1` - same as they are on the master branch. `v1alpha2` seems quite outdated to me.

File: Makefile [hunk 5]
```diff
@@ -278,6 +287,12 @@ generate-go-kubeadm-control-plane: $(CONTROLLER_GEN) $(CONVERSION_GEN) ## Runs G
 		--output-file-base=zz_generated.conversion \
 		--go-header-file=./hack/boilerplate/boilerplate.generatego.txt
 
+.PHONY: generate-go-operator
+generate-go-operator: $(CONTROLLER_GEN) ## Runs Go related generate targets for the operator
```
Comment (MarcelMue) [inline Makefile:291]: Running this updates a line in `exp/operator/api/v1alpha4/zz_generated.deepcopy.go` - pretty minor (just a comment) but would be nice to rerun in this PR.

**Review by MarcelMue — COMMENTED** [review]
Some comments, nothing critical. The man make commands seem sound to me.

File: exp/operator/config/default/manager_image_patch.yaml (hunk 2 above)
Comment (alexander-demichev) [reply exp/operator/config/default/manager_image_patch.yaml:10]: I'll remove this, it came from my gce config

File: Makefile [hunk 6]
```diff
@@ -87,6 +87,10 @@ KUBEADM_BOOTSTRAP_CONTROLLER_IMG ?= $(REGISTRY)/$(KUBEADM_BOOTSTRAP_IMAGE_NAME)
 KUBEADM_CONTROL_PLANE_IMAGE_NAME ?= kubeadm-control-plane-controller
 KUBEADM_CONTROL_PLANE_CONTROLLER_IMG ?= $(REGISTRY)/$(KUBEADM_CONTROL_PLANE_IMAGE_NAME)
 
+# operator
+OPERATOR_IMAGE_NAME ?= operator-controller
```
Comment (fabriziopandini) [inline Makefile:91]: `operator-controller`, uhmm ...
what about `cluster-api-operator`? this IMO will makes a nice pair with `cluster-api-controller`

File: Makefile [hunk 7]
```diff
@@ -87,6 +87,10 @@ KUBEADM_BOOTSTRAP_CONTROLLER_IMG ?= $(REGISTRY)/$(KUBEADM_BOOTSTRAP_IMAGE_NAME)
 KUBEADM_CONTROL_PLANE_IMAGE_NAME ?= kubeadm-control-plane-controller
 KUBEADM_CONTROL_PLANE_CONTROLLER_IMG ?= $(REGISTRY)/$(KUBEADM_CONTROL_PLANE_IMAGE_NAME)
 
+# operator
```
Comment (fabriziopandini) [inline Makefile:90]: background question:
given this is an experiment, should we host those targets in the main Makefile or have a separated Makefile like we are doing for CAPD?

File: Makefile [hunk 8]
```diff
@@ -468,6 +513,26 @@ release: clean-release ## Builds and push container images using the latest git
...
+	$(MAKE) set-manifest-image \
+		MANIFEST_IMG=$(PROD_REGISTRY)/$(KUBEADM_CONTROL_PLANE_IMAGE_NAME) MANIFEST_TAG=$(RELEASE_TAG) \
+		TARGET_RESOURCE="./controlplane/kubeadm/config/manager/manager_image_patch.yaml"
+	# Set the operator image to the production bucket.
+	$(MAKE) set-manifest-image \
+		MANIFEST_IMG=$(PROD_REGISTRY)/$(OPERATOR_IMAGE_NAME) MANIFEST_TAG=$(RELEASE_TAG) \
+		TARGET_RESOURCE="./exp/operator/config/default/manager_image_patch.yaml"
+	$(MAKE) set-manifest-pull-policy PULL_POLICY=IfNotPresent TARGET_RESOURCE="./config/manager/manager_pull_policy.yaml"
+	$(MAKE) set-manifest-pull-policy PULL_POLICY=IfNotPresent TARGET_RESOURCE="./bootstrap/kubeadm/config/manager/manager_pull_policy.yaml"
+	$(MAKE) set-manifest-pull-policy PULL_POLICY=IfNotPresent TARGET_RESOURCE="./controlplane/kubeadm/config/manager/manager_pull_policy.yaml"
```
Comment (fabriziopandini) [inline Makefile:534]: If I'm not wrong those changes are unrelated to this PR.
If they are required let's apply them on main and then cherry-pick (or rebase) on this branch

File: exp/operator/config/default/kustomization.yaml (hunk 3 above)
Comment (fabriziopandini) [reply exp/operator/config/default/kustomization.yaml:13]: We are already applying clusterctl labels to all the providers, and to cert-manager, I think this makes things consistent and it will make easier for clusterctl to manage the operator if required (e.g for upgrades)

File: exp/operator/controllers/coreprovider_controller.go (hunk 1 above)
Comment (fabriziopandini) [reply exp/operator/controllers/coreprovider_controller.go:40]: +1 to drop if not required

File: operator/config/crd/bases/operator.cluster.x-k8s.io_bootstrapproviders.yaml [hunk 9]
```diff
@@ -0,0 +1,46 @@
+
+---
```
Comment (fabriziopandini) [inline operator/config/crd/bases/operator.cluster.x-k8s.io_bootstrapproviders.yaml:2]: Am I wrong or all the operator files are expect to go under `exp/operator`?
Also this seems a duplicate of exp/operator/config/crd/bases/operator.cluster.x-k8s.io_bootstrapproviders.yaml

Comment (fabriziopandini) [reply operator/config/crd/bases/operator.cluster.x-k8s.io_bootstrapproviders.yaml:2]: Same for all the files under `operator/`

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
This PR contains far more changes than "Add make commands for CAPI operator"... 

For sake of clarity I suggest to change the PR description/ review the linked issue, or eventually to break down these 20 commits in separated PRs (some changes are unrelated: Makefile, yaml files under /config, main).

Also, it PR seems this PR is addressing some points in https://github.com/kubernetes-sigs/cluster-api/issues/4246.
Is it possible to document which point are we addressing on the isse/on the PR description?

File: exp/operator/controllers/coreprovider_controller.go (hunk 1 above)
Comment (alexander-demichev) [reply exp/operator/controllers/coreprovider_controller.go]: done

File: exp/operator/config/certmanager/certificate.yaml (hunk 4 above)
Comment (alexander-demichev) [reply exp/operator/config/certmanager/certificate.yaml]: right, the same story as with logger, can be done in a follow-up PR, I'll remove this blank space fix too. This is also tracked here https://github.com/kubernetes-sigs/cluster-api/issues/4246

File: operator/config/crd/bases/operator.cluster.x-k8s.io_bootstrapproviders.yaml (hunk 9 above)
Comment (alexander-demichev) [reply operator/config/crd/bases/operator.cluster.x-k8s.io_bootstrapproviders.yaml:2]: I removed them all, thanks.

**Comment by alexander-demichev** [comment]
@fabriziopandini @MarcelMue I tried to address your comment and removed some commits from the PR. All removed changes are tracked here https://github.com/kubernetes-sigs/cluster-api/issues/4246 and will be located in follow-up PRse

File: operator/config/crd/bases/operator.cluster.x-k8s.io_bootstrapproviders.yaml (hunk 9 above)
Comment (MarcelMue) [reply operator/config/crd/bases/operator.cluster.x-k8s.io_bootstrapproviders.yaml:2]: I still see the files under `operator/...` instead of `exp/operator/...`

File: Makefile [hunk 10]
```diff
@@ -87,6 +87,10 @@ KUBEADM_BOOTSTRAP_CONTROLLER_IMG ?= $(REGISTRY)/$(KUBEADM_BOOTSTRAP_IMAGE_NAME)
 KUBEADM_CONTROL_PLANE_IMAGE_NAME ?= kubeadm-control-plane-controller
 KUBEADM_CONTROL_PLANE_CONTROLLER_IMG ?= $(REGISTRY)/$(KUBEADM_CONTROL_PLANE_IMAGE_NAME)
 
+# operator
+OPERATOR_IMAGE_NAME ?= cluster-api-operator-controller
```
Comment (fabriziopandini) [inline Makefile:91]:
```suggestion
OPERATOR_IMAGE_NAME ?= cluster-api-operator
```
What about dropping controller?

**Review by fabriziopandini — COMMENTED** [review]
It seems some comment are still not addressed (@MarcelMue pointed out one of them)
Also, if I'm not wrong, background questions/general remarks are still to be addressed.

File: Makefile (hunk 7 above)
Comment (alexander-demichev) [reply Makefile:90]: yes, this is a good suggestion. I'm happy to do this in a follow-up PR.

File: operator/config/crd/bases/operator.cluster.x-k8s.io_bootstrapproviders.yaml (hunk 9 above)
Comment (alexander-demichev) [reply operator/config/crd/bases/operator.cluster.x-k8s.io_bootstrapproviders.yaml:2]: oh, sorry. I might've pushed some outdated changes.

**Comment by alexander-demichev** [comment]
/retitle Operator: Add makefile commands and make repo more CAPI consistent

**Comment by alexander-demichev** [comment]
List of issues https://github.com/kubernetes-sigs/cluster-api/issues/4246 addressed in this PR : 
- avoid a new .gitignore and rely on the top level one (e.g. we are already doing this for KCP, CABPK)
- avoid a new docker file and rely on the top level one (e.g. we are already doing this for KCP, CABPK)
- target a new namespace (capi-operator-system)
- add the clusterctl.cluster.x-k8s.io/core="operator" label to all the objects
- prefix all the object names with capi-operator
- introduce manager_image_patch.yaml manager_pull_policy.yaml patches (see consistent dev UX)
- add the node-role.kubernetes.io/master toleration
- make roles consistent with other operators (1 file with all the roles)
- add context to reconcile() arguments

**Review by MarcelMue — APPROVED** [review]
This is fine with me, we can iterate. Let's wait for Fabrizio to check again tomorrow.

File: exp/operator/controllers/coreprovider_controller.go [hunk 11]
```diff
@@ -37,7 +37,7 @@ type CoreProviderReconciler struct {
 // +kubebuilder:rbac:groups=operator.cluster.x-k8s.io,resources=coreproviders,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operator.cluster.x-k8s.io,resources=coreproviders/status,verbs=get;update;patch
 
-func (r *CoreProviderReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```
Comment (vincepri) [inline exp/operator/controllers/coreprovider_controller.go:40]: Unrelated change?

File: Makefile (hunk 7 above)
Comment (vincepri) [reply Makefile:90]: Could we have those changes in this PR? It seems appropriate given that we're just introducing this.

File: exp/operator/controllers/coreprovider_controller.go (hunk 11 above)
Comment (alexander-demichev) [reply exp/operator/controllers/coreprovider_controller.go:40]: Reconciler interface requires context so it won't build without this change:
`
Reconcile(context.Context, Request) (Result, error)
`

File: Makefile [hunk 12]
```diff
@@ -237,7 +230,7 @@ generate-go: $(GOBINDATA) ## Runs Go related generate targets
 	$(MAKE) generate-go-core
 	$(MAKE) generate-go-kubeadm-bootstrap
 	$(MAKE) generate-go-kubeadm-control-plane
-	$(MAKE) generate-go-operator
+	$(MAKE) -C $(OPERATOR_DIR) generate-go-operator
```
Comment (fabriziopandini) [inline Makefile:233]: I think we should not call the operator targets by default.
Let's call them explicitly, like we are doing for CAPD

File: Makefile [hunk 13]
```diff
@@ -415,9 +390,9 @@ docker-build-kubeadm-control-plane: ## Build the docker image for kubeadm contro
 
 .PHONY: docker-build-operator
 docker-build-operator: ## Build the docker image for management cluster operator
-	DOCKER_BUILDKIT=1 docker build --build-arg goproxy=$(GOPROXY) --build-arg ARCH=$(ARCH) --build-arg package=./exp/operator --build-arg ldflags="$(LDFLAGS)" . -t $(OPERATOR_CONTROLLER_IMG)-$(ARCH):$(TAG)
-	$(MAKE) set-manifest-image MANIFEST_IMG=$(OPERATOR_CONTROLLER_IMG)-$(ARCH) MANIFEST_TAG=$(TAG) TARGET_RESOURCE="./exp/operator/config/default/manager_image_patch.yaml"
-	$(MAKE) set-manifest-pull-policy TARGET_RESOURCE="./exp/operator/config/default/manager_pull_policy.yaml"
+	$(MAKE) -C $(OPERATOR_DIR) docker-build-operator
```
Comment (fabriziopandini) [inline Makefile:393]: this target should go away from the main docker file

File: Makefile [hunk 14]
```diff
@@ -474,11 +449,8 @@ docker-push-kubeadm-control-plane-manifest: ## Push the fat manifest docker imag
 
 .PHONY: docker-push-operator-manifest
 docker-push-operator-manifest: ## Push the fat manifest docker image for the operator image.
-	## Minimum docker version 18.06.0 is required for creating and pushing manifest images.
-	docker manifest create --amend $(OPERATOR_CONTROLLER_IMG):$(TAG) $(shell echo $(ALL_ARCH) | sed -e "s~[^ ]*~$(OPERATOR_CONTROLLER_IMG)\-&:$(TAG)~g")
-	@for arch in $(ALL_ARCH); do docker manifest annotate --arch $${arch} ${OPERATOR_CONTROLLER_IMG}:${TAG} ${OPERATOR_CONTROLLER_IMG}-$${arch}:${TAG}; done
-	docker manifest push --purge $(OPERATOR_CONTROLLER_IMG):$(TAG)
-	$(MAKE) set-manifest-image MANIFEST_IMG=$(OPERATOR_CONTROLLER_IMG) MANIFEST_TAG=$(TAG) TARGET_RESOURCE="./exp/operator/config/default/manager_image_patch.yaml"
+	$(MAKE) -C "./$(OPERATOR_DIR)/" docker-push-operator-manifest
```
Comment (fabriziopandini) [inline Makefile:452]: this target should go away from the main docker file

File: Makefile [hunk 15]
```diff
@@ -315,7 +302,7 @@ generate-manifests: ## Generate manifests e.g. CRD, RBAC etc.
 	$(MAKE) generate-core-manifests
 	$(MAKE) generate-kubeadm-bootstrap-manifests
 	$(MAKE) generate-kubeadm-control-plane-manifests
-	$(MAKE) generate-operator-manifests
+	$(MAKE) -C $(OPERATOR_DIR) generate-operator-manifests
```
Comment (fabriziopandini) [inline Makefile:305]: same

File: Makefile [hunk 16]
```diff
@@ -546,7 +518,7 @@ release-manifests: $(RELEASE_DIR) $(KUSTOMIZE) ## Builds the manifests to publis
 	# Build control-plane-components.
 	$(KUSTOMIZE) build controlplane/kubeadm/config/default > $(RELEASE_DIR)/control-plane-components.yaml
 	# Build operator components.
-	$(KUSTOMIZE) build exp/operator/config/default > $(RELEASE_DIR)/operator-components.yaml
+	$(KUSTOMIZE) build $(OPERATOR_DIR)/config/default > $(RELEASE_DIR)/operator-components.yaml
```
Comment (fabriziopandini) [inline Makefile:521]: This should go in the operator make file

File: Makefile [hunk 17]
```diff
@@ -531,11 +503,11 @@ manifest-modification: # Set the manifest images to the staging/production bucke
 		TARGET_RESOURCE="./controlplane/kubeadm/config/default/manager_image_patch.yaml"
 	$(MAKE) set-manifest-image \
 		MANIFEST_IMG=$(PROD_REGISTRY)/$(OPERATOR_IMAGE_NAME) MANIFEST_TAG=$(RELEASE_TAG) \
-		TARGET_RESOURCE="./exp/operator/config/default/manager_image_patch.yaml"
+		TARGET_RESOURCE="./$(OPERATOR_DIR)/config/default/manager_image_patch.yaml"
 	$(MAKE) set-manifest-pull-policy PULL_POLICY=IfNotPresent TARGET_RESOURCE="./config/default/manager_pull_policy.yaml"
 	$(MAKE) set-manifest-pull-policy PULL_POLICY=IfNotPresent TARGET_RESOURCE="./bootstrap/kubeadm/config/default/manager_pull_policy.yaml"
 	$(MAKE) set-manifest-pull-policy PULL_POLICY=IfNotPresent TARGET_RESOURCE="./controlplane/kubeadm/config/default/manager_pull_policy.yaml"
-	$(MAKE) set-manifest-pull-policy PULL_POLICY=IfNotPresent TARGET_RESOURCE="./exp/operator/config/default/manager_pull_policy.yaml"
+	$(MAKE) set-manifest-pull-policy PULL_POLICY=IfNotPresent TARGET_RESOURCE="./$(OPERATOR_DIR)/config/default/manager_pull_policy.yaml"
```
Comment (fabriziopandini) [inline Makefile:510]: This should go in the operator make file

File: Makefile [hunk 18]
```diff
@@ -531,11 +503,11 @@ manifest-modification: # Set the manifest images to the staging/production bucke
 		TARGET_RESOURCE="./controlplane/kubeadm/config/default/manager_image_patch.yaml"
 	$(MAKE) set-manifest-image \
 		MANIFEST_IMG=$(PROD_REGISTRY)/$(OPERATOR_IMAGE_NAME) MANIFEST_TAG=$(RELEASE_TAG) \
-		TARGET_RESOURCE="./exp/operator/config/default/manager_image_patch.yaml"
+		TARGET_RESOURCE="./$(OPERATOR_DIR)/config/default/manager_image_patch.yaml"
```
Comment (fabriziopandini) [inline Makefile:506]: This should go in the operator make file

File: Makefile (hunk 12 above)
Comment (alexander-demichev) [reply Makefile:233]: I thought the operator is a bit different, don't we plan to ship it the same way as `kubeadm` stuff?

Comment (Danil-Grigorev) [reply Makefile:233]: This makes sense IMO, and it will help identify early issues during manifest generation, when generated resources will overlap between CAPI and CAPIO. CI should handle this part and prevent from PRs being merged if this task fails.

File: Makefile [hunk 19]
```diff
@@ -405,6 +415,7 @@ docker-push-all: $(addprefix docker-push-,$(ALL_ARCH))
 	$(MAKE) docker-push-core-manifest
 	$(MAKE) docker-push-kubeadm-bootstrap-manifest
 	$(MAKE) docker-push-kubeadm-control-plane-manifest
+	$(MAKE) docker-push-operator-manifest
```
Comment (Danil-Grigorev) [inline Makefile:418]: I would argue that this task should be independent from core Makefile. Until the operator is experimental, default push command should accidentally create new images in release.

**Comment by alexander-demichev** [comment]
A small recap: 
- Added a separate Makefile for the operator(similar to CAPD) and separated the PR into meaningful commits.
- Added context to `Reconcile()` arguments and removed zap logger from test. Without these 2 changes, the project fails to build.
- Couple of other minor changes.

**Review by MarcelMue — APPROVED** [review]
I would like to go forward with this for now. We will have an easier time iterating then.

**Comment by fabriziopandini** [comment]
+1 to get this moving.
Please squash commits

**Comment by MarcelMue** [comment]
Good to go from my side

/LGTM
